File: spring-boot-commons-util/src/test/java/com/github/springframework/boot/commons/util/NumbersTest.java
package com.github.springframework.boot.commons.util;

import com.github.springframework.boot.commons.util.base.Numbers;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.math.BigDecimal;
import java.math.BigInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class NumbersTest {

    @Test
    void testNewInstance() {
        UnsupportedOperationException e = assertThrows(UnsupportedOperationException.class, Numbers::new);
        assertEquals("Utility class should not be instantiated", e.getMessage());
    }

    @ParameterizedTest
    @CsvSource({", 1024, 1024", "1024, 2048, 1024"})
    void testDefaultIfNullInteger(Integer number, int defaultValue, int expected) {
        assertEquals(expected, Numbers.defaultIfNull(number, defaultValue));
    }

    @ParameterizedTest
    @CsvSource({", 1024, 1024", "1024, 2048, 1024"})
    void testDefaultIfNullLong(Long number, long defaultValue, long expected) {
        assertEquals(expected, Numbers.defaultIfNull(number, defaultValue));
    }

    @ParameterizedTest
    @CsvSource({", 1024.0, 1024.0", "1024.0, 2048.0, 1024.0"})
    void testDefaultIfNullFloat(Float number, float defaultValue, float expected) {
        assertEquals(expected, Numbers.defaultIfNull(number, defaultValue));
    }

    @ParameterizedTest
    @CsvSource({", 1024.0, 1024.0", "1024.0, 2048.0, 1024.0"})
    void testDefaultIfNullDouble(Double number, double defaultValue, double expected) {
        assertEquals(expected, Numbers.defaultIfNull(number, defaultValue));
    }

    @ParameterizedTest
    @CsvSource({", 1024, 1024", "1024, 2048, 1024"})
    void testDefaultIfNullBigInteger(BigInteger number, BigInteger defaultValue, BigInteger expected) {
        assertEquals(expected, Numbers.defaultIfNull(number, defaultValue));
    }

    @ParameterizedTest
    @CsvSource({", 1024.0, 1024.0", "1024.0, 2048.0, 1024.0"})
    void testDefaultIfNullBigDecimal(BigDecimal number, BigDecimal defaultValue, BigDecimal expected) {
        assertEquals(expected, Numbers.defaultIfNull(number, defaultValue));
    }

    @ParameterizedTest
    @CsvSource({", 0", "1024, 1024"})
    void testZeroIfNullInteger(Integer number, int expected) {
        assertEquals(expected, Numbers.zeroIfNull(number));
    }

    @ParameterizedTest
    @CsvSource({", 0", "1024, 1024"})
    void testZeroIfNullLong(Long number, long expected) {
        assertEquals(expected, Numbers.zeroIfNull(number));
    }

    @ParameterizedTest
    @CsvSource({", 0.0", "1024.0, 1024.0"})
    void testZeroIfNullFloat(Float number, float expected) {
        assertEquals(expected, Numbers.zeroIfNull(number));
    }

    @ParameterizedTest
    @CsvSource({", 0.0", "1024.0, 1024.0"})
    void testZeroIfNullDouble(Double number, double expected) {
        assertEquals(expected, Numbers.zeroIfNull(number));
    }

    @ParameterizedTest
    @CsvSource({", 0", "1024, 1024"})
    void testZeroIfNullBigInteger(BigInteger number, BigInteger expected) {
        assertEquals(expected, Numbers.zeroIfNull(number));
    }

    @ParameterizedTest
    @CsvSource({", 0", "1024, 1024"})
    void testZeroIfNullBigDecimal(BigDecimal number, BigDecimal expected) {
        assertEquals(expected, Numbers.zeroIfNull(number));
    }

    @ParameterizedTest
    @CsvSource({
        ", false", "'', false", "+, false", "-, false", "123, true", "123.45, true",
        "+123.45, true", "-123.45, true", "123+45, false", "123-45, false", "123.45.67, false"
    })
    void testCanBeConvertedToBigDecimal(String number, boolean expected) {
        assertEquals(expected, Numbers.canBeConvertedToBigDecimal(number));
    }

}

File: spring-boot-commons-base/src/test/java/com/github/awesome/springboot/commons/base/NumbersTest.kt
package com.github.awesome.springboot.commons.base

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.math.BigDecimal
import java.math.RoundingMode
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class NumbersTest {

    @Test
    fun testNewInstance() {
        val e = assertThrows<UnsupportedOperationException> {
            Numbers()
        }
        assertEquals("Utility class should not be instantiated", e.message)
    }

    @Test
    fun testDefaultIfNull() {
        assertEquals(1, Numbers.defaultIfNull(null, 1))
        assertEquals(BigDecimal.ZERO, Numbers.defaultIfNull(null, BigDecimal.ZERO))
        assertEquals(1, Numbers.defaultIfNull(1, 0))
        assertEquals(BigDecimal.ONE, Numbers.defaultIfNull(BigDecimal.ONE, BigDecimal.ZERO))
    }

    @Test
    fun testIsZero() {
        assertTrue { Numbers.isZero(BigDecimal("0.00")) }
        assertTrue { Numbers.isZero(BigDecimal("0.0")) }
        assertTrue { Numbers.isZero(BigDecimal("0")) }
        assertTrue { Numbers.isZero(BigDecimal.ZERO) }
        assertFalse { Numbers.isZero(BigDecimal.ONE) }
        assertFalse { Numbers.isZero(null) }
    }

    @Test
    fun testIsPositive() {
        assertTrue { Numbers.isPositive(BigDecimal.ONE) }
        assertFalse { Numbers.isPositive(null) }
        assertFalse { Numbers.isPositive(BigDecimal.valueOf(-1)) }
    }

    @Test
    fun testIsNegative() {
        assertTrue { Numbers.isNegative(BigDecimal.valueOf(-1)) }
        assertFalse { Numbers.isNegative(null) }
        assertFalse { Numbers.isNegative(BigDecimal.ONE) }
    }

    @Test
    fun testToPlainString() {
        assertEquals("1", Numbers.toPlainString(BigDecimal.ONE, 0, RoundingMode.HALF_UP))
        assertEquals("1.0", Numbers.toPlainString(BigDecimal.ONE, 1, RoundingMode.HALF_UP))
        assertEquals("1.00", Numbers.toPlainString(BigDecimal.ONE, 2, RoundingMode.HALF_UP))
        assertEquals("1.23", Numbers.toPlainString(BigDecimal.valueOf(1.23), 2, RoundingMode.HALF_UP))
        assertEquals("1.23", Numbers.toPlainString(BigDecimal.valueOf(1.234), 2, RoundingMode.HALF_UP))
        assertEquals("1.24", Numbers.toPlainString(BigDecimal.valueOf(1.235), 2, RoundingMode.HALF_UP))
    }

    @Test
    fun testToScaledPlainString() {
        assertEquals("1", Numbers.toScaledPlainString(BigDecimal.ONE, 0))
        assertEquals("1.0", Numbers.toScaledPlainString(BigDecimal.ONE, 1))
        assertEquals("1.00", Numbers.toScaledPlainString(BigDecimal.ONE, 2))
        assertEquals("1.23", Numbers.toScaledPlainString(BigDecimal.valueOf(1.23), 2))
        assertEquals("1.23", Numbers.toScaledPlainString(BigDecimal.valueOf(1.234), 2))
        assertEquals("1.24", Numbers.toScaledPlainString(BigDecimal.valueOf(1.235), 2))
    }

    @Test
    fun testIsBigDecimalConvertible() {
        assertTrue { Numbers.isBigDecimalConvertible("1") }
        assertTrue { Numbers.isBigDecimalConvertible("1.0") }
        assertTrue { Numbers.isBigDecimalConvertible("1.00") }
        assertTrue { Numbers.isBigDecimalConvertible("1.23") }
        assertTrue { Numbers.isBigDecimalConvertible("1.234") }
        assertTrue { Numbers.isBigDecimalConvertible("1.235") }
        assertFalse { Numbers.isBigDecimalConvertible("1.23.45") }
        assertFalse { Numbers.isBigDecimalConvertible(Strings.EMPTY) }
        assertFalse { Numbers.isBigDecimalConvertible(null) }
    }

}
